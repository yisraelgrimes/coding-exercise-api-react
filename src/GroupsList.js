/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect } from "react";
import _ from "lodash";
import axios from "axios";
import { Table, List } from "semantic-ui-react";
import { catchErrors } from "./utils/catchErrors";

const apiUrl = "http://127.0.0.1:8000/api/groups";


export default function GroupsList({jestData}) {
    const [containerState, setContainerState] = useState({
        column: null,
        data: [],
        direction: null,
    });

    // Get the list of groups
    useEffect(() => {
        async function getData() {
            try {
                const response = await axios.get(apiUrl);
                setContainerState((prevState) => ({
                    ...prevState,
                    data: response.data.data,
                }));
            } catch (error) {
                catchErrors(error);
            }
        }
        getData();
    }, []);


    const handleSort = (clickedColumn) => {
        const { column, data, direction } = containerState;
        if (column !== clickedColumn) {
            setContainerState((prevState) => ({
                ...prevState,
                column: clickedColumn,
                data: _.sortBy(data, [clickedColumn]),
                direction: "ascending",
            }));
            return;
        }

        setContainerState((prevState) => ({
            ...prevState,
            data: data.reverse(),
            direction: direction === "ascending" ? "descending" : "ascending",
        }));
    };
    const { column, data, direction } = containerState;

    // If we are using Jest, use the 'jestData' prop to load test data
    let groupsData = data;
    if (jestData) {
        groupsData = jestData;
    }

    return (
        <Table sortable celled padded>
            <Table.Header>
                <Table.Row>
                    <Table.HeaderCell
                        sorted={column === "name" ? direction : null}
                        onClick={() => handleSort("name")}
                        singleLine
                    >
                        Group Name
                    </Table.HeaderCell>
                    <Table.HeaderCell
                        sorted={column === "people" ? direction : null}
                        onClick={() => handleSort("people")}
                        singleLine
                    >
                        People
                    </Table.HeaderCell>
                </Table.Row>
            </Table.Header>

            {groupsData &&
            <Table.Body>
                {/* Map through groups */}
                {groupsData.map(g => {
                    return (
                        <Table.Row key={g.id}>
                            <Table.Cell singleLine>{ g.group_name }</Table.Cell>

                            {/* Map through a group's people and show active */}
                            <Table.Cell singleLine>
                                {g.people.length > 0 && (
                                    <List>
                                        {g.people.map(p => {
                                            if (p.status === "active") {
                                                return (
                                                    <List.Item key={p.id}>
                                                        {p.first_name} {p.last_name}
                                                    </List.Item>
                                                );
                                            }
                                            return null;
                                        })}
                                    </List>
                                )}
                            </Table.Cell>
                        </Table.Row>
                    );
                })}
            </Table.Body>
            }
        </Table>
    );
}
