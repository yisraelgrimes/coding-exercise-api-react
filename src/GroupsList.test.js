/* eslint-disable no-undef */
import React from "react";
import { shallow } from "enzyme";
import GroupsList from "./GroupsList";

let wrapper;

const data = [{
    "id": 1,
    "group_name": "Volunteers",
    "people": [{
        "id": 132,
        "first_name": "Macie",
        "last_name": "Emmerich",
        "email_address": "cremin.marjory@hotmail.com",
        "status": "active",
        "group_id": 1,
        "updated_at": "2019-07-20 22:05:47",
        "created_at": "2019-07-20 22:05:47",
    }],
    "updated_at": "2019-07-21 22:05:47",
    "created_at": "2019-07-21 22:05:47",
}];

describe("<GroupsList />", () => {

    beforeAll(() => {
        wrapper = shallow(<GroupsList jestData={data} />);
    });

    test("should match the snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });
});
