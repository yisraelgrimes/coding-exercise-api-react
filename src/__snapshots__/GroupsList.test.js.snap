// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupsList /> should match the snapshot 1`] = `
<Table
  as="table"
  celled={true}
  padded={true}
  sortable={true}
>
  <TableHeader
    as="thead"
  >
    <TableRow
      as="tr"
      cellAs="td"
    >
      <TableHeaderCell
        as="th"
        onClick={[Function]}
        singleLine={true}
        sorted={null}
      >
        Group Name
      </TableHeaderCell>
      <TableHeaderCell
        as="th"
        onClick={[Function]}
        singleLine={true}
        sorted={null}
      >
        People
      </TableHeaderCell>
    </TableRow>
  </TableHeader>
  <TableBody
    as="tbody"
  >
    <TableRow
      as="tr"
      cellAs="td"
      key="1"
    >
      <TableCell
        as="td"
        singleLine={true}
      >
        Volunteers
      </TableCell>
      <TableCell
        as="td"
        singleLine={true}
      >
        <List>
          <ListItem
            key="132"
          >
            Macie
             
            Emmerich
          </ListItem>
        </List>
      </TableCell>
    </TableRow>
  </TableBody>
</Table>
`;
